File: 209.minimum-size-subarray-sum.go
/*
 * @lc app=leetcode id=209 lang=golang
 *
 * [209] Minimum Size Subarray Sum
 */

// @lc code=start
func minSubArrayLen(target int, nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	res, sum, i, j := len(nums)+1, nums[0], 0, 0

	for j < len(nums) {
		if sum < target && j < len(nums) {
			j++

			if j == len(nums){
				continue
			}

			sum += nums[j]
		} 

		if sum >= target {
			sum -= nums[i]
			res = min(res, j-i+1)
			i++
		}
	}

	if res == len(nums)+1 {
		return 0
	}

	return res
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
// @lc code=end

File: 209.minimum-size-subarray-sum.cpp
/*
 * @lc app=leetcode id=209 lang=cpp
 *
 * [209] Minimum Size Subarray Sum
 */

// @lc code=start
class Solution {
 public:
  int minSubArrayLen(int target, vector<int>& nums) {
    int l = 0;

    int ans = INT_MAX;
    int sum = 0;
    // slide window
    for (int r = 0; r < nums.size(); ++r) {
      sum += nums[r];
      while (sum >= target) {
        ans = min(ans, (r - l + 1));
        sum -= nums[l++];
      }
    }

    return ans == INT_MAX ? 0 : ans;
  }
};
// @lc code=end

